File: frontend/src/types/usage.ts
export type UsageSourceType = 'assistant' | 'workflow';
export type UsageSourceFilter = UsageSourceType | 'all';
export type UsageRange = '24h' | '7d' | '30d';
export type AdminUsageRange = UsageRange;
export type UsageWindow = '5m' | '60m';

export interface UsageMetrics {
  logical_request_count: number;
  provider_call_count: number;
  successful_provider_call_count: number;
  total_tokens: number;
  total_cost_usd: number;
  error_rate: number;
  fallback_rate: number;
  p95_latency_ms: number | null;
}

export interface UsageOverviewResponse {
  range: string;
  assistant: UsageMetrics;
  workflow: UsageMetrics;
  all: UsageMetrics;
}

export interface UsageLivePoint {
  ts: string;
  logical_requests: number;
  provider_calls: number;
  successful_provider_calls: number;
  total_tokens: number;
  total_cost_usd: number;
  error_count: number;
  fallback_count: number;
}

export interface UsageLiveResponse {
  window: string;
  points: UsageLivePoint[];
  summary: UsageMetrics;
}

export interface UsageTimeseriesPoint {
  ts: string;
  assistant_calls: number;
  workflow_calls: number;
  assistant_tokens: number;
  workflow_tokens: number;
  assistant_cost_usd: number;
  workflow_cost_usd: number;
}

export interface UsageTimeseriesResponse {
  range: string;
  source: UsageSourceFilter;
  points: UsageTimeseriesPoint[];
}

export interface ModelBreakdownItem {
  provider: string;
  model: string;
  provider_call_count: number;
  successful_provider_call_count: number;
  total_tokens: number;
  total_cost_usd: number;
  success_rate: number;
}

export interface ModelBreakdownResponse {
  range: string;
  source: UsageSourceFilter;
  items: ModelBreakdownItem[];
}

export interface RecentCallItem {
  id: string;
  request_id: string;
  source_type: UsageSourceType;
  source_subtype: string;
  provider: string;
  model: string;
  node_id: string | null;
  status: string;
  is_fallback: boolean;
  latency_ms: number | null;
  total_tokens: number;
  cost_amount_usd: number;
  started_at: string;
}

export interface RecentCallsResponse {
  calls: RecentCallItem[];
}

export interface CostSplitItem {
  source_type: UsageSourceType;
  provider_call_count: number;
  total_tokens: number;
  total_cost_usd: number;
}

export interface CostSplitResponse {
  range: string;
  items: CostSplitItem[];
}

File: backend/app/models/usage.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


UsageSourceType = Literal["assistant", "workflow"]
UsageSourceFilter = Literal["assistant", "workflow", "all"]
UsageRange = Literal["24h", "7d", "30d"]
UsageWindow = Literal["5m", "60m"]


class UsageMetrics(BaseModel):
    logical_request_count: int = 0
    provider_call_count: int = 0
    successful_provider_call_count: int = 0
    total_tokens: int = 0
    total_cost_usd: float = 0.0
    error_rate: float = 0.0
    fallback_rate: float = 0.0
    p95_latency_ms: int | None = None


class UsageOverviewResponse(BaseModel):
    range: str
    assistant: UsageMetrics
    workflow: UsageMetrics
    all: UsageMetrics


class UsageLivePoint(BaseModel):
    ts: datetime
    logical_requests: int = 0
    provider_calls: int = 0
    successful_provider_calls: int = 0
    total_tokens: int = 0
    total_cost_usd: float = 0.0
    error_count: int = 0
    fallback_count: int = 0


class UsageLiveResponse(BaseModel):
    window: str
    points: list[UsageLivePoint] = Field(default_factory=list)
    summary: UsageMetrics


class UsageTimeseriesPoint(BaseModel):
    ts: str
    assistant_calls: int = 0
    workflow_calls: int = 0
    assistant_tokens: int = 0
    workflow_tokens: int = 0
    assistant_cost_usd: float = 0.0
    workflow_cost_usd: float = 0.0


class UsageTimeseriesResponse(BaseModel):
    range: str
    source: UsageSourceFilter
    points: list[UsageTimeseriesPoint] = Field(default_factory=list)


class ModelBreakdownItem(BaseModel):
    provider: str
    model: str
    provider_call_count: int = 0
    successful_provider_call_count: int = 0
    total_tokens: int = 0
    total_cost_usd: float = 0.0
    success_rate: float = 0.0


class ModelBreakdownResponse(BaseModel):
    range: str
    source: UsageSourceFilter
    items: list[ModelBreakdownItem] = Field(default_factory=list)


class RecentCallItem(BaseModel):
    id: str
    request_id: str
    source_type: UsageSourceType
    source_subtype: str
    provider: str
    model: str
    node_id: str | None = None
    status: str
    is_fallback: bool
    latency_ms: int | None = None
    total_tokens: int = 0
    cost_amount_usd: float = 0.0
    started_at: datetime


class RecentCallsResponse(BaseModel):
    calls: list[RecentCallItem] = Field(default_factory=list)


class CostSplitItem(BaseModel):
    source_type: UsageSourceType
    provider_call_count: int = 0
    total_tokens: int = 0
    total_cost_usd: float = 0.0


class CostSplitResponse(BaseModel):
    range: str
    items: list[CostSplitItem] = Field(default_factory=list)
